### PR TITLE
chore(streaming): rewrite lookup v2 for lower latency

### PR DESCRIPTION
**Pull request by skyzh**
Signed-off-by: Alex Chi <iskyzh@gmail.com>

## What's changed and what's your intention?

This PR contains a thorough rewrite of the lookup executor. Improvements include:

* As less dynamic dispatch as possible.
* Less waiting when looking up the current epoch. Messages will now be immediately available after arrangement is ready.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

### CODE REVIEW

File: src/stream/src/executor_v2/error.rs [hunk 1]
```diff
@@ -47,6 +49,9 @@ pub enum StreamExecutorError {
 
     #[error("Channel `{0}` closed")]
     ChannelClosed(String),
+
+    #[error("Failed to align barrier: expected {0:?} but got {1:?}")]
+    AlignBarrier(Box<Barrier>, Box<Barrier>),
```
Comment (BugenZhao) [inline src/stream/src/executor_v2/error.rs:54]: Could you please help to use this error in `Merge`?

https://github.com/singularity-data/risingwave/blob/9499dc7434d884961ef178115470a104ef24415e/src/stream/src/executor_v2/merge.rs#L182-L187

**Review by BugenZhao — APPROVED** [review]

File: src/stream/src/executor_v2/error.rs (hunk 1 above)
Comment (skyzh) [reply src/stream/src/executor_v2/error.rs:54]: Next time for sure 🤪

Comment (skyzh) [reply src/stream/src/executor_v2/error.rs:54]: And I think it's better to use two different errors at two different places.

Comment (BugenZhao) [reply src/stream/src/executor_v2/error.rs:54]: They're both for aligning barriers. 😭

**Pull request merged**